lib/system/mmdisp.nim

```diff
@@ -214,175 +210,125 @@ elif defined(gogc):
              "[GC] number of frees: " & $(mstats.nfree) & "\n" &
              "[GC] heap objects: " & $(mstats.heap_objects) & "\n" &
-             "[GC] numgc: " & $(mstats.numgc) & "\n" &
-             "[GC] enablegc: " & $(mstats.enablegc) & "\n" &
-             "[GC] debuggc: " & $(mstats.debuggc) & "\n" &
-             "[GC] total pause time [ms]: " & $(mstats.pause_total_ns div 1000_000)
+             "[GC] number of completed GC cycles: " & $(mstats.numgc) & "\n" &
+             "[GC] total GC pause time [ms]: " & $(mstats.pause_total_ns div 1000_000)
 
   proc getOccupiedMem(): int =
-    var mstats: goMStats
-    goRuntime_ReadMemStats(addr mstats)
+    var mstats = goMemStats()
     result = int(mstats.alloc)
 
   proc getFreeMem(): int =
-    var mstats: goMStats
-    goRuntime_ReadMemStats(addr mstats)
+    var mstats = goMemStats()
     result = int(mstats.sys - mstats.alloc)
 
   proc getTotalMem(): int =
-    var mstats: goMStats
-    goRuntime_ReadMemStats(addr mstats)
+    var mstats = goMemStats()
     result = int(mstats.sys)
 
   proc nimGC_setStackBottom(theStackBottom: pointer) = discard
 
   proc alloc(size: Natural): pointer =
-    result = c_malloc(size)
-    if result == nil: raiseOutOfMem()
+    result = goMalloc(size.uint)
 
   proc alloc0(size: Natural): pointer =
-    result = alloc(size)
-    zeroMem(result, size)
+    result = goMalloc(size.uint)
 
   proc realloc(p: pointer, newsize: Natural): pointer =
-    result = c_realloc(p, newsize)
-    if result == nil: raiseOutOfMem()
+    raise newException(Exception, "not implemented")
 
-  proc dealloc(p: pointer) = c_free(p)
+  proc dealloc(p: pointer) =
+    discard
 
   proc allocShared(size: Natural): pointer =
-    result = c_malloc(size)
-    if result == nil: raiseOutOfMem()
+    result = alloc(size)
 
   proc allocShared0(size: Natural): pointer =
-    result = alloc(size)
-    zeroMem(result, size)
+    result = alloc0(size)
 
   proc reallocShared(p: pointer, newsize: Natural): pointer =
-    result = c_realloc(p, newsize)
-    if result == nil: raiseOutOfMem()
+    result = realloc(p, newsize)
 
-  proc deallocShared(p: pointer) = c_free(p)
+  proc deallocShared(p: pointer) = dealloc(p)
 
   when hasThreadSupport:
     proc getFreeSharedMem(): int = discard
     proc getTotalSharedMem(): int = discard
     proc getOccupiedSharedMem(): int = discard
 
-  const goFlagNoZero: uint32 = 1 shl 3
-  proc goRuntimeMallocGC(size: uint, typ: uint, flag: uint32): pointer {.
-    importc: "runtime_mallocgc", dynlib: goLib.}
-
-  proc goSetFinalizer(obj: pointer, f: pointer) {.
-    importc: "set_finalizer", codegenDecl: "$1 $2$3 __asm__ (\"main.Set_finalizer\");\n$1 $2$3",
-    dynlib: goLib.}
-
   proc newObj(typ: PNimType, size: int): pointer {.compilerproc.} =
-    result = goRuntimeMallocGC(roundup(size, sizeof(pointer)).uint, 0.uint, 0.uint32)
+    writebarrierptr(addr(result), goMalloc(size.uint))
```

Comment:
https://github.com/gcc-mirror/gcc/blob/f5edd321a43ac1c07145d6759cabddedd5c4f110/libgo/go/runtime/mbarrier.go#L119 : 

_"One might be tempted to ignore the write barrier if slot points
into to the stack. Don't do it! Mark termination only re-scans
frames that have potentially been active since the concurrent scan,
so it depends on write barriers to track changes to pointers in
stack frames that have not been active."_

They explain here how they gave up a traditional at-stop-the-world-time stack scanning GC for a fancy tricolor one that avoids that: https://github.com/golang/proposal/blob/master/design/17503-eliminate-rescan.md

Reply:
Ok but then I hope you found all the locations where the C codegen should produce a write barrier for stack slots, because it wasn't designed with this in mind.

Reply:
I did my best to spot them, but I'll keep the stack in mind if I ever see memory corruption from premature garbage collection.